dotnet/WPF/VisualLayerIntegration/VisualLayerIntegration/CompositionHost.cs

```diff
@@ -31,42 +32,29 @@
 
 namespace VisualLayerIntegration
 {
-      class CompositionHost : HwndHost
+    sealed class CompositionHost : HwndHost
     {
-        object dispatcherQueue;
-        int hostHeight, hostWidth;
-        double hostDpiX, hostDpiY;
-        private ICompositionTarget compositionTarget;
-        internal const int
-            WS_CHILD = 0x40000000,
-            WS_VISIBLE = 0x10000000,
-            LBS_NOTIFY = 0x00000001,
-            HOST_ID = 0x00000002,
-            LISTBOX_ID = 0x00000001,
-            WS_VSCROLL = 0x00200000,
-            WS_BORDER = 0x00800000;
-
-
-        public Visual Child
-        {
-            set
-            {
-                if (Compositor == null)
-                {
-                    InitComposition(hwndHost);
-                }
-                compositionTarget.Root = value;
-            }
-        }
+        private object _dispatcherQueue;
+        private ICompositionTarget _compositionTarget;
+        private List<IDisposable> registeredDisposables = new List<IDisposable>();
+       
         public IntPtr hwndHost { get; private set; }
         public Compositor Compositor { get; private set; }
 
-        public CompositionHost(double height, double width, double dpiX, double dpiY)
+        public event EventHandler<HwndMouseEventArgs> MouseLClick;
+        public event EventHandler<HwndMouseEventArgs> MouseMoved;
+        public event EventHandler<InvalidateDrawingEventArgs> InvalidateDrawing;
+
+
+        public CompositionHost() {   }
+
+        public void SetChild(Visual v)
         {
-            hostHeight = (int)height;
-            hostWidth = (int)width;
-            hostDpiX = (double)dpiX;
-            hostDpiY = (double)dpiY;
+            if (Compositor == null)
+            {
+                InitializeComposition(hwndHost);
+            }
```

Comment:
As far as I can tell, the Compositor will never be null because it was used to create the visual. If it was null then you'd end up with 2 compositors - one that the Visual was created on, and the other that you created. That won't work.
A comment on this method explaining what has to happen before this is called would help a lot - clearly there is some protocol where the CompositionHost is instantiated but then BuildWindowCore has to happen before it's possible to use the Compositor property. It's hard to know when it's safe to get the Compositor without a little more documentation.